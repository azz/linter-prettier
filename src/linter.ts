import { LinterAdapter, LinterFactory, LinterProvider } from "@linter/core";
import { getSupportInfo } from "prettier";
import { lint } from "prettylint";
import { logger, getLogLevel } from "./logger";
import {
  prettylintResultToFormatOutput,
  prettylintResultToLintOutput,
} from "./utils";

const logLevel = getLogLevel();
logger.debug(`Setting log level to "${logLevel}"`);
logger.setLevel(logLevel);

const linter: LinterAdapter = {
  async format({ filePath, text }) {
    logger.debug("Running format");
    logger.debug(`filePath: ${filePath}`);
    logger.debug(`text: ${text}`);

    logger.debug("Running prettylint");
    const result = await lint(filePath, text);
    const formatOutput = prettylintResultToFormatOutput(result);
    logger.debug("Format done");
    return formatOutput;
  },

  async lint({ filePath, text }) {
    logger.debug("Running lint");
    logger.debug(`filePath: ${filePath}`);
    logger.debug(`text: ${text}`);

    logger.debug("Running prettylint");
    const result = await lint(filePath, text);
    const lintOutput = prettylintResultToLintOutput(result);
    logger.debug("Lint done");
    return lintOutput;
  },
};

const linterFactory: LinterFactory = () => linter;

const linterProvider: LinterProvider = {
  factory: linterFactory,
  name: "prettier",
  supportedExtensions: getSupportInfo().languages.reduce(
    (accumulator, { extensions }) => accumulator.concat(extensions),
    [] as string[],
  ),
};

export { linter, linterFactory, linterProvider };
