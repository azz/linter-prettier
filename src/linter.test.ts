import fs from "jest-plugin-fs";
import { sep as separator } from "path";
import { linter, linterFactory } from "./linter";

jest.mock("fs", () => {
  const realFs = require.requireActual("fs");
  const mockFs = require("jest-plugin-fs/mock");
  const { ufs } = require("unionfs");
  ufs.use(realFs).use(mockFs);
  return ufs;
});

beforeAll(() => {
  fs.root = process.cwd();
});

afterEach(() => {
  fs.restore();
  jest.resetModules();
});

describe("Format", () => {
  test("with text and filePath", async () => {
    const linter = await linterFactory();

    expect(
      await linter.format({ filePath: "foo.js", text: "var bar = 'baz'" }),
    ).toMatchSnapshot();
  });

  test("with already formatted text", async () => {
    const linter = await linterFactory();

    expect(
      await linter.format({ filePath: "foo.js", text: 'var foo = "bar";' }),
    ).toMatchSnapshot();
  });

  test("with ignored file", async () => {
    fs.mock({
      ".prettierignore": "foo.js",
    });

    const linter = await linterFactory();

    expect(
      await linter.format({ filePath: "foo.js", text: "var bar = 'baz'" }),
    ).toMatchSnapshot();
  });
});

describe("Lint", () => {
  test("with text and filePath", async () => {
    const linter = await linterFactory();

    expect(
      await linter.lint({ filePath: "foo.js", text: "var bar = 'baz'" }),
    ).toMatchSnapshot();
  });

  test("with syntax error", async () => {
    const linter = await linterFactory();

    expect(
      await linter.lint({ filePath: "foo.js", text: "var foo ==== 'bar'" }),
    ).toMatchSnapshot();
  });
});
