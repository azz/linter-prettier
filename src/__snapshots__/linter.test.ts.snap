// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Format with already formatted text 1`] = `
Object {
  "errorCount": 0,
  "filePath": "foo.js",
  "messages": Array [
    Object {
      "column": 16,
      "line": 0,
      "message": "Insert \`⏎\`",
      "ruleId": "prettier/prettier",
      "severity": 1,
    },
  ],
  "output": "var foo = \\"bar\\";
",
  "warningCount": 1,
}
`;

exports[`Format with ignored file 1`] = `
Object {
  "errorCount": 0,
  "filePath": "foo.js",
  "messages": Array [
    Object {
      "column": 10,
      "line": 0,
      "message": "Replace \`'baz'\` with \`\\"baz\\";⏎\`",
      "ruleId": "prettier/prettier",
      "severity": 1,
    },
  ],
  "output": "var bar = \\"baz\\";
",
  "warningCount": 1,
}
`;

exports[`Format with text and filePath 1`] = `
Object {
  "errorCount": 0,
  "filePath": "foo.js",
  "messages": Array [
    Object {
      "column": 10,
      "line": 0,
      "message": "Replace \`'baz'\` with \`\\"baz\\";⏎\`",
      "ruleId": "prettier/prettier",
      "severity": 1,
    },
  ],
  "output": "var bar = \\"baz\\";
",
  "warningCount": 1,
}
`;

exports[`Lint with syntax error 1`] = `
Object {
  "errorCount": 1,
  "filePath": "foo.js",
  "messages": Array [
    Object {
      "column": 9,
      "line": 1,
      "message": "Unexpected token, expected \\";\\" (1:9)",
      "ruleId": "prettier/prettier",
      "severity": 2,
    },
  ],
  "warningCount": 0,
}
`;

exports[`Lint with text and filePath 1`] = `
Object {
  "errorCount": 0,
  "filePath": "foo.js",
  "messages": Array [
    Object {
      "column": 10,
      "line": 0,
      "message": "Replace \`'baz'\` with \`\\"baz\\";⏎\`",
      "ruleId": "prettier/prettier",
      "severity": 1,
    },
  ],
  "warningCount": 1,
}
`;
